### ISSUE TITLE
Better lazy-download handling

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#16456]
Lazy-download is specified in the site local config handling by enabling it as well as specifying a list of eligible directories that can be used to store the temporary file. At the moment, the behavior of the system is to look to see if any of the eligible directories is both local to the machine and has sufficient storage space. The first directory (based on the order given by the user) which satisfies both criteria is the one chosen. If no directory matches that criteria then lazy-download is silently ignored and the file just read directly. The auto fallback to reading directly caused serious inefficiencies in the Tier 0.

I propose that we change the algorithm to be

1. choose the first eligible directory which is both local and has sufficient storage, if no such directory exists
1. choose the first eligible non-local directory which has sufficient storage, if non exist
1. based on a configuration parameter, either print a warning and fallback to reading the file directly or thrown an exception

Allowing non-local directories (if the user has specified one in their configuration) handles the use case where a site might be reading the file from the WAN but wants to cache it on a LAN file system.

In order to avoid changing the behavior of sites that are using lazy-download, I propose that the parameter which governs 'what to do when there is no useable directory' should default to 'print warning and read directly'. I propose the parameter be named: `throwExceptionIfCannotLazyDownload` and default to `false`.

**Comment by Dr15Jones** [comment]
assign core

**Comment by bouril** [comment]
Sounds fine to me. The Lustre caching case for reading files over the WAN is covered. Actually, we have observed better read rates over Lustre compared to local disks.

**Comment by bbockelm** [comment]
Is this a PSet parameter or something from the `SiteLocalConfigService`?

**Comment by Dr15Jones** [comment]
@bbockelm are you asking if the proposed new parameter `throwExceptionIfCannotLazyDownload` would be added to `SiteLocalConfigService`? If that is your question then the answer is "yes".

**Comment by bbockelm** [comment]
Ugh.  I asked an ambiguous question, as `SiteLocalConfigService` takes input from both `site-local-config.xml` and the PSet.  Let me try again:

Will this be set via `site-local-config.xml`, the PSet, or both?

**Comment by Dr15Jones** [comment]
I would intend it to be both just like all other parameters of `SiteLocalConfigService`.

**Comment by smuzaffar** [comment]
I do not see any ref to `throwExceptionIfCannotLazyDownload` so I guess this is not implemented yet

**Comment by makortel** [comment]
> I guess this is not implemented yet

Correct
